### DISCUSSION TITLE
Search and Replace Question

### DISCUSSION THREAD

**Discussion by cwebpro** [sqlitebrowser/sqlitebrowser#3088] [Q&A] [answered]
Just a quick question on the replace functionality for DB Browser.
I'm trying to use the "Replace Text in Cells" button while looking at a text column in a table.
I want to do a replace on just the text in one column (multiple cells) not the whole table.  I'm not seeing a way to do that and just want to make sure I'm not missing something (If I run a replace to replace an _  with a blank space, it does it for all the columns in the table not just what I selected).  Is there a good video on using DB Browser for SQLite I can watch that you would recommend?
Thanks, Matthew

**Comment by chrisjlocke** [comment, marked as answer]
Haven't used the search and replace before, but yes, it appears it runs through the whole database, not just one cell.
There is a 'next match' and 'previous match' to help you locate the ones you need, but thats still quite manual, and not helpful if you only want to change one cell and you have fifty...

However, DB4S allows SQL to be used quite easily.
Search and replace is a powerful tool, so ensure you have a database backup.  SQL isn't that forgiving... DB4S does have an 'undo' button (well, more specifically a 'write changes' button) but once you click that and realise it's replaced too much, it'll be too late.

So you can execute this sql:

    update table
    set field=replace(field,'text to find','text to replace with')

This only replaces text in one field.  However, it'll replace all instances it finds - you can't pick 'n choose.  Here's a quick video.

https://user-images.githubusercontent.com/7872218/177490645-1a82edbd-32f4-44a9-9101-d9108a366dfb.mp4

So in that instance, it replaces all occurrences of 'lamb' with 'chicken'.  If a field contained 'Mary lamb had a lamb little lamb' it would replace all of them.  If you only wanted the third one, then you'd have to change the search term or use regular expressions ... yell if you want to go down that rabbit hole of fun.

In the video I then added another column to show how you can refine the search and replace even more granular, to specific records

    update table
    set field=replace(field,'text to find','text to replace with')
    where a_field = some-value

You can also combine the search and replace

    update table
    set field=replace(field,'text to find','text to replace with'),
    field=replace(field,'more text to find','text to replace with');

So in my video, if I replaced 'lamb' with 'chicken' and 'Mary' with 'Sue', I'd get this

![image](https://user-images.githubusercontent.com/7872218/177491957-9f57be91-0c9f-4fa3-b261-a1ce2eb42c0f.png)

(restricted to one field...)

Hope this makes some kind of sense.

**Reply by chrisjlocke** [reply]
![image](https://user-images.githubusercontent.com/7872218/177492729-f74196c2-c9ac-40b2-bab0-b6c23d90429a.png)

In my table, there is no collation set (how it matches when I tell it to find something) so the above searches are case-sensitive.  Searching for 'LAMB' will not find 'lamb'.

**Reply by cwebpro** [reply]
Thank you Chris for the great reply and help.
That worked great for me.   I guess I need  to find a good tutorial on SQLite's SQL  (I've used some Oracle in the past but am new to SQLite).

Thanks again for the help ...
Matthew

**Comment by chrisjlocke** [comment]
Happy to help where I can.  SQL is quite a small language.  
I'd recommend the w3schools tutorials
https://www.w3schools.com/sql/

The examples they give isn't great, but it gives you the idea.

![image](https://user-images.githubusercontent.com/7872218/177590738-9aa92991-bf41-4710-a3e6-88207f305346.png)

If you're creating your own database, I'd recommend setting the collation to 'nocase', so that 'find all records that contain fred' will find 'Fred' and 'FRED'.

**Reply by chrisjlocke** [reply]
Another handy tip of the day, once you strart playing with views.  Views allow you to combine different tables into a sort of 'virtual table' (you can't add or edit records** in a view - they're handy for just ... viewing ... tables)
** you sort of can, but its a fudge.  Don't go down that road.  You can't.  end of.

Go to the pragma tab, and type **pragma alter_table_legacy=1**
This allows you to edit tables that have views connected to them ... otherwise DB4S moans.  You have to enter that each time you open a database.  There are project files, but ... well, don't go their either. ;)

**Reply by chrisjlocke** [reply]
I should do a beginners guied in the wiki!
Oh, that reminds me - we have a wiki.
https://github.com/sqlitebrowser/sqlitebrowser/wiki

Its not extensive, but again, yell back with 'How do I?' and we can fill it in.

Welcome to the world of DB4S and SQLite!

**Reply by chrisjlocke** [reply]
Also, we have https://dbhub.io - this is a website/portal/sharing thing where you can upload databases and have different versions, ala github.  This is where the 'remote' tab comes into play in DB4S.
